pkg/kubelet/apis/device-plugin/v1alpha1/api.proto

```diff
@@ -0,0 +1,118 @@
+option (gogoproto.goproto_getters_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_unrecognized_all) = false;
+
+
+// Registration is the service advertised by the Kubelet
+// Only when Kubelet answers with a success code to a Register Request
+// may Device Plugins start their service
+service Registration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+}
+
+// DevicePlugin is the service advertised by Device Plugins
+service DevicePlugin {
+	// ListWatch returns a stream of List of Devices
+	// Whenever a Device state change or a Device disapears, ListWatch
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+
+	// Allocate is called during container creation so that the Device
+	// Plugin can run device specific operations and instruct Kubelet
+	// of the steps to make the Device available in the container
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string endpoint = 2;
+	// Name of the Device advertised same as the one in the Device struct
+	string resourceName = 3;
+}
+
+message RegisterResponse {
+	// Minimum version the Kubelet API supports.
+	bool success = 1;
+	// Kubelet fills this field if it encounters any errors
+	// during the registration process
+	string error = 2;
+}
+
+// Allocate is called during container creation so that the Device
+// Plugin can run device specific operations on the Devices sent
+message AllocateRequest {
+	repeated Device devices = 1;
+}
+
+message AllocateResponse {
+	repeated DeviceRuntimeSpec spec = 1;
+}
+
+// ListWatch returns a stream of List of Devices
+// Whenever a Device state change or a Device disapears, ListWatch
+// returns the new list
+message ListWatchResponse {
+	repeated Device devices = 1;
+}
+
+// The list of be added to the CRI spec
+message DeviceRuntimeSpec {
+	// List of environment variable to set in the container.
+	repeated KeyValue envs = 1;
+	// Mounts for the container.
+	repeated Mount mounts = 2;
+}
+
+// Mount specifies a host volume to mount into a container.
+// where device library or tools are installed on host and container
+message Mount {
+	// Path of the mount on the host.
+	string host_path = 1;
+	// Path of the mount within the container.
+	string mount_path = 2;
+	// If set, the mount is read-only.
+	bool read_only = 3;
+}
+
+message KeyValue {
+	string key = 1;
+	string value = 2;
+}
+
+/* E.g:
+* struct Device {
+*    Kind: "nvidia-gpu",
+*    ID: "GPU-fef8089b-4820-abfc-e83e-94318197576e",
+*    State: "Healthy",
+*    Properties: {
+*        "Family": "Pascal",
+*        "Memory": "4G",
+*        "ECC"   : "True",
+*    }
+*}
+*/
+message Device {
+	string resourceName = 1;
```

Comment:
Can we omit resourceName here given that it is already included in RegisterRequest? If we include this field in both places, Kubelet side needs to detect and handle the case that a device plugin sends a Device struct with non-existing or conflicting resourceName. To detect this, Kubelet side needs to maintain some kind of mapping from device plugin connection to resource_name. So keeping this field in Device struct doesn't seem to provide extra value but just adds more error cases to worry about. If we don't include this field here, the API contract is simpler. Each device is uniquely identified by its id within a device plugin. Kubelet side maintains some kind mapping from device plugin connection to resourceName, but only needs to detect possible conflicting resource_name during registration.


[10:12] 
Similarly, during Allocate, Kubelet only needs to send device ids to device plugin. It doesn't need to send the whole Device struct.